Simplify Bolt11 Payments and Remove Redundant Code

shaavan opened the pull request:
resolves #3262
builds on: #3342 

### Summary  
This PR introduces `pay_for_bolt11_invoice`, simplifying how Bolt11 payments are processed by allowing direct invoice-based payments while still enabling custom routing parameters. With this change, the redundant `bolt11_payment.rs` utility functions are no longer needed and have been removed.  

### Why This Change?  
- Previously, paying a Bolt11 invoice required manually extracting `payment_id`, `payment_hash`, and other parameters before passing them to `send_payment`.  
- `pay_for_bolt11_invoice` introduces a more streamlined approach, mirroring the existing convenience of Bolt12 payments.  
- With this function now handling parameter extraction within `ChannelManager`, `bolt11_payment.rs` is redundant and can be safely removed.  

### Changes  
- **Introduced `pay_for_bolt11_invoice`**:  
  - Takes a Bolt11 invoice directly.  
  - Allows users to set custom routing parameters via `RouteParametersConfig`.  
- **Removed `bolt11_payment.rs`**:  
  - All relevant logic is now handled within `ChannelManager`.  
  - Removed associated test cases for redundant utility functions.  

### Impact  
- Reduces complexity for Bolt11 payments.  
- Cleans up redundant code, making the implementation more maintainable.

shaavan commented:
Updated from [pr3617.01](https://github.com/shaavan/rust-lightning/commits/pr3617.01) to [pr3617.02](https://github.com/shaavan/rust-lightning/commits/pr3617.02) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3617.01..pr3617.02)):

Changes:
1. Rebase on main.

shaavan commented:
Updated from [pr3617.02](https://github.com/shaavan/rust-lightning/commits/pr3617.02) to [pr3617.03](https://github.com/shaavan/rust-lightning/commits/pr3617.03) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3617.02..pr3617.03)):
Addressed @TheBlueMatt comments

Changes:
1. Introduce payment_id as parameter, and update documentation

shaavan commented on lightning/src/ln/outbound_payment.rs:
```diff
@@ -847,6 +850,49 @@ impl OutboundPayments {
+		SP: Fn(SendAlongPathArgs) -> Result<(), APIError>,
+	{
+		let payment_hash = PaymentHash((*invoice.payment_hash()).to_byte_array());
+		let payment_id = PaymentId(payment_hash.0);
```
Thanks for the pointer! Updated in [pr3617.03](https://github.com/shaavan/rust-lightning/commits/pr3617.03)

shaavan commented on lightning/src/ln/outbound_payment.rs:
```diff
@@ -847,6 +850,49 @@ impl OutboundPayments {
+
+		let amount = match (invoice.amount_milli_satoshis(), amount_msats) {
+			(Some(amt), None) | (None, Some(amt)) => amt,
+			(None, None) | (Some(_), Some(_)) => return Err(RetryableSendFailure::DuplicatePayment), // change error properly
```
Oops! Thanks for catching it! 🙃
Updated in [pr3617.03](https://github.com/shaavan/rust-lightning/commits/pr3617.03)

jkczyz commented on lightning/src/events/mod.rs:
```diff
@@ -1702,7 +1707,9 @@ impl Writeable for Event {
 					Some(PaymentFailureReason::BlindedPathCreationFailed) =>
-						&Some(PaymentFailureReason::RouteNotFound)
+						&Some(PaymentFailureReason::RouteNotFound),
+					Some(PaymentFailureReason::InvalidAmount) =>
+						&Some(PaymentFailureReason::InvalidAmount)
```
Won't this break downgrades? We need to map to a variant available prior to version 0.0.124. I guess it is never serialized in an event, and instead return directly? I wonder if we should make a `Bolt11PaymentError` similar to `Bolt12PaymentError` instead.

jkczyz commented on lightning/src/ln/bolt11_payment.rs:
```diff
@@ -1,216 +0,0 @@
-	}
-
-	#[test]
-	fn payment_metadata_end_to_end() {
```
Seems like we should keep this test (moving it), but use the new API and also cover the zero-amount cases.

shaavan replied:
Makes sense! I moved the test into our codebase in [pr3617.04](https://github.com/shaavan/rust-lightning/commits/pr3617.04)
Also, introduced a variation on it with `invoice_with_no_amount` case.
Thanks so much for the pointer!

jkczyz replied:
Could you also test the failure cases? (i.e., when returning `Bolt11PaymentError::InvalidAmount`)?

shaavan replied:
Done! [pr3617.05](https://github.com/shaavan/rust-lightning/commits/pr3617.05)
Thanks! ✨

shaavan commented:
Updated from [pr3617.03](https://github.com/shaavan/rust-lightning/commits/pr3617.03) to [pr3617.04](https://github.com/shaavan/rust-lightning/commits/pr3617.04) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3617.03..pr3617.04)):
Addressed @jkczyz comments

Changes:
1. Introduce new variant Bolt11PaymentError, to not break downgrades.
2. Move the relevant test from bolt11_payment.rs, to the codebase, to keep the test suite exhaustive.

jkczyz commented on lightning/src/ln/invoice_utils.rs:
```diff
@@ -808,6 +804,123 @@ mod test {
 	}
 
+	#[test]
+	fn payment_metadata_end_to_end_for_invoice_with_amount() {
```
@TheBlueMatt Not sure where you prefer these tests go. I guess we can move them once / if `invoice_utils.rs` is removed. I'm somewhat inclined to rename `bolt11_payment.rs` to `bolt11_payment_tests.rs`.

shaavan replied:
Done! [pr3617.06](https://github.com/shaavan/rust-lightning/commits/pr3617.06)
I have renamed the file to bolt11_payment_tests.rs, and moved the new introduced function within it.
Thanks a lot for the suggestion! ✨

shaavan commented:
Updated from [pr3617.05](https://github.com/shaavan/rust-lightning/commits/pr3617.05) to [pr3617.06](https://github.com/shaavan/rust-lightning/commits/pr3617.06) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3617.05..pr3617.06)):
Addressed @jkczyz comment

Changes:
1. Renamed the `bolt11_payment.rs` -> `bolt11_payment_tests.rs`
2. Move the introduce tests to `bolt11_payment_tests.rs`

shaavan commented:
Updated from [pr3617.06](https://github.com/shaavan/rust-lightning/commits/pr3617.06) to [pr3617.07](https://github.com/shaavan/rust-lightning/commits/pr3617.07) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3617.06..pr3617.07)):
Addressed @TheBlueMatt comments

Changes:
1. Rebase on main to resolve merge conflicts

jkczyz reviewed (commented):
Looks like `bolt11_payment_tests.rs` is using spaces instead of tabs

shaavan commented:
Updated from [pr3617.07](https://github.com/shaavan/rust-lightning/commits/pr3617.07) to [pr3617.08](https://github.com/shaavan/rust-lightning/commits/pr3617.08) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3617.07..pr3617.08)):

Changes:
1. Fixed rust formatting for `bolt11_payment_tests.rs`
